docs/architecture/deep-dives/state/index.md

```diff
@@ -0,0 +1,379 @@
+
+### `SingleUserState<T>`
+
+`SingleUserState<T>` behaves very similarly to `GlobalState<T>` where neither will react to active
+user changes and you instead give it the user you want it to care about up front, which is publicly
+exposed as a `readonly` member.
+
+Updates to `SingleUserState` or `ActiveUserState` handling the same `KeyDefinition` will cause each
+other to emit on their `state$` observables if the `userId` handled by the `SingleUserState` happens
+to be active at the time of the update.
+
+## Migrating
+
+Migrating data to state providers is incredibly similar to migrating data in general. You create
+your own class that extends `Migrator<From, To>`. That will require you to implement your own
+`migrate(migrationHelper: MigrationHelper)` method. `MigrationHelper` already includes methods like
+`get` and `set` for getting and settings value to storage by their string key. There are also
+methods for getting and setting using your `KeyDefinition` or `KeyDefinitionLike` object to and from
+user and global state. An example of how you might use these new helpers is below:
+
+```typescript
+type ExpectedGlobalState = { myGlobalData: string };
+
+type ExpectedAccountState = { myUserData: string };
+
+const MY_GLOBAL_KEY_DEFINITION: KeyDefinitionLike = {
+  stateDefinition: { name: "myState" },
+  key: "myGlobalKey",
+};
+const MY_USER_KEY_DEFINITION: KeyDefinitionLike = {
+  stateDefinition: { name: "myState" },
+  key: "myUserKey",
+};
+
+export class MoveToStateProvider extends Migrator<10, 11> {
+  async migrate(migrationHelper: MigrationHelper): Promise<void> {
+    const existingGlobalData = await migrationHelper.get<ExpectedGlobalState>("global");
+
+    await migrationHelper.setGlobal(MY_GLOBAL_KEY_DEFINITION, {
+      myGlobalData: existingGlobalData.myGlobalData,
+    });
+
+    const updateAccount = async (userId: string, account: ExpectedAccountState) => {
+      await migrationHelper.setUser(MY_USER_KEY_DEFINITION, {
+        myUserData: account.myUserData,
+      });
+    };
+
+    const accounts = await migrationHelper.getAccounts<ExpectedAccountState>();
+
+    await Promise.all(accounts.map(({ userId, account }) => updateAccount(userId, account)));
+  }
+}
+```
+
+:::note
+
+`getAccounts` only gets data from the legacy account object that was used in `StateService`. As data
+gets migrated off of that account object the response from `getAccounts`, which returns a record
+where the key will be a users id and the value being the legacy account object.
+
+_TODO: Implement method just for user ids_
+
+:::
+
+### Example PRs
+
+_TODO: Include PR's_
+
+## Testing
+
+Testing business logic with data and observables can sometimes be cumbersome, to help make that a
+little easier, there are a suite of helpful "fakes" that can be used instead of traditional "mocks".
+Now instead of calling `mock<StateProvider>()` into your service you can instead use
+`new FakeStateProvider()`.
+
+_TODO: Refine user story_
+
+## Advanced usage
```

Comment:
🤔 How do I manage state that is encrypted on disk but plaintext in memory? I imagine I'd want to use a disk and a memory store and wire them together in sequence?

1. RxJs application sequence updates memory store.
2. Memory store triggers Rxjs encryption sequence.
3. Rxjs encryption sequence updates disk store.

All of the back-and-forth with application logic would end up going back and forth with the memory store. The disk store would be used for loading, then disconnect. Later, when an update comes through it'd load long enough to save the update then go quiet again. I imagine that means that it's timeout could be kept pretty short--perhaps on the order of milliseconds, or perhaps even immediately?

Reply:
We are working on this exact problem! @MGibson1 has a draft PR right now where we are working on a good solution to these types of problems. bitwarden/clients#7290. So far everything we've thrown out is very inline with what you've mentioned here. Except for one thing, I think right now the decrypted state is going to be read-only. If you want to update it we are going to make you manually re-encrypt it and then call `update` to your disk state. But I've mentioned Matt above so he can maybe take this for some ideas!

Reply:
Oh, I always figured the encryption and decryption was manually performed, since there's no way the storage layer could differentiate.

Now that I think about it, though, there's some interesting possibilities that come to mind if you were to divide the state into encrypted parts and plaintext parts. You could sync the encrypted parts using content-addressable storage and k-anonymity tokens stored with the plaintext. That's probably way out of scope of this, though.

Reply:
The decryption _is_ still performed by code written by each owner of that state, you just have to give us that function and we decide _when_ to run it. Basically we will run it for you when you have updated the encrypted data AND someone is subscribed to the derived data (someone cares ❤️). 

We didn't do anything near as fancy as you described and we have merged the derived state PR but if you want to have a chat about the pros of doing it like you described I'd be down to listen. As of now we do have to pass around user key access and crypto services every pretty much which I am definitely not in love with.

Reply:
I think we kind of did have this conversation, though without the CAS portions. If I have a chance to experiment with it as part of Sends, then I'll let you know about the branch so we can figure out if it's something that we should polish and lift.